### ISSUE TITLE
bug(linux): Deb package "ibus-keyman" breaks apt and can't be installed or removed

### ISSUE THREAD

**Issue by dboehmer** [keymanapp/keyman#3902]
**Describe the bug**
I try to help a linguist installing Keyman on Ubuntu 18.04 LTS (still supported, all updates installed). The Keyman packages are in a broken state, can’t be removed and block **any** package to be installed.

**To Reproduce**
I couldn’t inspect this remotely and actually went to install a VM for myself to reproduce this.

1. Install Ubuntu 18.04 LTS x86-64
2. `sudo add-apt-repository ppa:keymanapp/keyman`
3. `sudo apt install keyman` (worked fine)
3. `sudo apt remove keyman` (worked)
4. `sudo apt remove ibus-keyman` (breaks)

Now I can’t fully remove Keyman. I think this is the actual cause of the problem.

```
daniel@daniel-Standard-PC-Q35-ICH9-2009:~$ sudo apt remove ibus-keyman 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages will be REMOVED:
  ibus-keyman
0 upgraded, 0 newly installed, 1 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 73,7 kB disk space will be freed.
Do you want to continue? [Y/n] 
(Reading database ... 203316 files and directories currently installed.)
Removing ibus-keyman (13.0.103-1~sil1~bionic) ...
Can't connect to IBus.
dpkg: error processing package ibus-keyman (--remove):
 installed ibus-keyman package post-removal script subprocess returned error exit status 1
Errors were encountered while processing:
 ibus-keyman
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

Now I also can’t install the `keyman` package again. I needed to add `python3-keyman-config` and `ibus-keyman` because they are dependencies and were explicitly removed before.

```
daniel@daniel-Standard-PC-Q35-ICH9-2009:~$ sudo apt install keyman python3-keyman-config ibus-keyman
Reading package lists... Done
Building dependency tree       
Reading state information... Done
ibus-keyman is already the newest version (13.0.103-1~sil1~bionic).
keyman is already the newest version (13.0.103-1~sil1~bionic).
python3-keyman-config is already the newest version (13.0.103-1~sil1~bionic).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
3 not fully installed or removed.
Need to get 0 B/21,4 kB of archives.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] 
dpkg: dependency problems prevent configuration of python3-keyman-config:
 python3-keyman-config depends on ibus-keyman; however:
  Package ibus-keyman is not installed.

dpkg: error processing package python3-keyman-config (--configure):
 dependency problems - leaving unconfigured
No apport report written because the error message indicates its a followup error from a previous failure.
                                                                                                          dpkg: dependency problems prevent configuration of keyman:
 keyman depends on python3-keyman-config; however:
  Package python3-keyman-config is not configured yet.

dpkg: error processing package keyman (--configure):
 dependency problems - leaving unconfigured
No apport report written because the error message indicates its a followup error from a previous failure.
                                                                                                          Errors were encountered while processing:
 python3-keyman-config
 keyman
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

Unfortunately this breaks the whole apt and blocks any package from being installed, e.g. `openssh-server`:

```
daniel@daniel-Standard-PC-Q35-ICH9-2009:~$ sudo apt install openssh-server
Reading package lists... Done
Building dependency tree       
Reading state information... Done
You might want to run 'apt --fix-broken install' to correct these.
The following packages have unmet dependencies:
 openssh-server : Depends: openssh-sftp-server but it is not going to be installed
                  Recommends: ncurses-term but it is not going to be installed
                  Recommends: ssh-import-id but it is not going to be installed
 python3-keyman-config : Depends: ibus-keyman but it is not going to be installed
E: Unmet dependencies. Try 'apt --fix-broken install' with no packages (or specify a solution).
```

The fix isn’t possible because `ibus-keyman` fails to complete installation *and* fails to uninstall.

**Expected behavior**
All packages from the Keyman PPA should be allowed to install or remove.

---

**Keyman Desktop, Keyman for macOS, Keyman for Linux:**
 - OS: Ubuntu 18.04
 - Keyman version: `13.0.103-1~sil1~bionic`
 - Target Application: next step
  
---
 
**Keyboard**
 - Keyboard name: IPA
 - Keyboard version: can’t tell
 - Language name: –

**Additional context**
When `apt` is back in a usable state we need to setup Keyman itself and open the virtual keyboard. When I started Keyman in my VM after first install, I could add the IPA keyboard but I don’t know yet where to find the virtual keyboard.

**Comment by ermshiperete** [comment]
So far I can't repro this problem. Works as expected for me. However, I notice that `ibus` doesn't seem to be running on your machine when it tries to ibus-keyman - which is strange.

Which UI do you use? Gnome? What flavour of Ubuntu?

Can you paste the output of the following commands: 

`im-config`

`dpkg -l | grep ibus`

`ps -ef | grep ibus`

**Comment by dboehmer** [comment]
> So far I can't repro this problem. Works as expected for me. However, I notice that `ibus` doesn't seem to be running on your machine when it tries to ibus-keyman - which is strange.

Hmm, I tried to describe my steps as precisely as possible. I have no clue of Keyman or ibus yet. Maybe I am missing a step that you consider obvious? I only installed packages and ran `keyman`.

> Which UI do you use? Gnome? What flavour of Ubuntu?

The linguist uses Kubuntu. I tried with standard Ubuntu. Both 18.04 LTS. I started under Gnome. Later I installed KDE Plasma but the packages were already in the described state.

> Can you paste the output of the following commands:
> 
> `im-config`

```
daniel@daniel-Standard-PC-Q35-ICH9-2009:~$ im-config

(zenity:2625): Gtk-WARNING **: 17:27:30.782: Theme parsing error: gtk.css:68:35: The style property GtkButton:child-displacement-x is deprecated and shouldn't be used anymore. It will be removed in a future version

(zenity:2625): Gtk-WARNING **: 17:27:30.783: Theme parsing error: gtk.css:69:35: The style property GtkButton:child-displacement-y is deprecated and shouldn't be used anymore. It will be removed in a future version
Gtk-Message: 17:27:30.863: GtkDialog mapped without a transient parent. This is discouraged.
```

A window *Input Method Configuration (im-config, ver. 0.34-1ubuntu1.3)* opened with the following in a textbox:

```
Current configuration for the input method:
 * Active configuration: missing (normally missing)
 * Normal automatic choice: ibus (normally ibus or fcitx or uim)
 * Override rule: zh_CN,fcitx:zh_TW,fcitx:zh_HK,fcitx:zh_SG,fcitx:ja_JP,fcitx:ko_KR,fcitx:vi_VN,fcitx
 * Current override choice:  (en_US)
 * Current automatic choice: ibus
 * Number of valid choices: 2 (normally 1)
The override rule is defined in /etc/default/im-config.
The configuration set by im-config is activated by re-starting X.
Explicit selection is not required to enable the automatic configuration if the active one is default/auto/cjkv/missing.
  Available input methods: ibus xim
Unless you really need them all, please make sure to install only one input method tool.
```

Then I cancelled the dialog because I have no idea what I'm doing there.

> `dpkg -l | grep ibus`

```
daniel@daniel-Standard-PC-Q35-ICH9-2009:~$ dpkg -l | grep ibus
ii  gir1.2-ibus-1.0:amd64                           1.5.17-3ubuntu5.3                                amd64        Intelligent Input Bus - introspection data
ii  ibus                                            1.5.17-3ubuntu5.3                                amd64        Intelligent Input Bus - core
ii  ibus-gtk:amd64                                  1.5.17-3ubuntu5.3                                amd64        Intelligent Input Bus - GTK+2 support
ii  ibus-gtk3:amd64                                 1.5.17-3ubuntu5.3                                amd64        Intelligent Input Bus - GTK+3 support
iH  ibus-keyman                                     13.0.103-1~sil1~bionic                           amd64        Keyman engine for IBus
ii  ibus-table                                      1.9.14-3                                         all          table engine for IBus
ii  libgusb2:amd64                                  0.2.11-1                                         amd64        GLib wrapper around libusb1
ii  libibus-1.0-5:amd64                             1.5.17-3ubuntu5.3                                amd64        Intelligent Input Bus - shared library
ii  libusageenvironment3:amd64                      2018.02.18-1                                     amd64        multimedia RTSP streaming library (UsageEnvironment classes)
ii  libusb-1.0-0:amd64                              2:1.0.21-2                                       amd64        userspace USB programming library
ii  libusbmuxd4:amd64                               1.1.0~git20171206.c724e70f-0.1                   amd64        USB multiplexor daemon for iPhone and iPod Touch devices - librar
```

> `ps -ef | grep ibus`

```
daniel@daniel-Standard-PC-Q35-ICH9-2009:~$ ps -ef | grep ibus
daniel    3760  1480  0 17:32 pts/0    00:00:00 grep --color=auto ibus
```

**Comment by ermshiperete** [comment]
@dboehmer Thanks! I'm still not able to exactly reproduce your problem, but I was able to get into a state similar to yours. I also found (and fixed) a problem in the uninstall scripts, so in the future this should work better.

To work around the problem you can try the following steps:

1. reboot your machine
2. after login in open a terminal window and run: `ps -C ibus-daemon -f`

This should output something like this:

```
UID        PID  PPID  C STIME TTY          TIME CMD
eberhard  2107  1172  0 18:18 tty1     00:00:00 ibus-daemon --xim --panel disabl
```

If you get only the header line (UID...) the ibus is not running. In this case please paste the output and I need to look further.

3. Otherwise run `sudo apt --fix-broken install`

This will remove the broken packages.

Afterwards you can reinstall keyman (or whatever).

**Comment by dboehmer** [comment]
> @dboehmer Thanks! I'm still not able to exactly reproduce your problem, but I was able to get into a state similar to yours. I also found (and fixed) a problem in the uninstall scripts, so in the future this should work better.

Thanks, I saw the commit and I’m glad you realized the problem.

> To work around the problem you can try the following steps:
> 
>     1. reboot your machine
>     2. after login in open a terminal window and run: `ps -C ibus-daemon -f`
>     ....

After reboot I was automatically logged into KDE Plasma. ibus was **not** running (only the header line).

I logged out and logged manually into "Ubuntu" (Gnome). With Gnome it seems to behave differently:

```
daniel@daniel-Standard-PC-Q35-ICH9-2009:~$ ps -C ibus-daemon -f
UID        PID  PPID  C STIME TTY          TIME CMD
gdm       1578  1544  0 00:17 tty1     00:00:00 ibus-daemon --xim --panel disabl
daniel    1875  1853  0 00:18 tty2     00:00:00 ibus-daemon --xim --panel disabl
```
 
> If you get only the header line (UID...) the ibus is not running. In this case please paste the output and I need to look further.
> 
>     1. Otherwise run `sudo apt --fix-broken install`

The command actually ran fine!

```
daniel@daniel-Standard-PC-Q35-ICH9-2009:~$ sudo apt install --fix-broken 
[sudo] password for daniel: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages were automatically installed and are no longer required:
  kmflcomp libblas3 libgfortran4 libkmflcomp0 libkmnkbp0-0 liblapack3
  libquadmath0 python3-bs4 python3-html5lib python3-lxml python3-magic
  python3-numpy python3-qrcode python3-requests-cache python3-webencodings
Use 'sudo apt autoremove' to remove them.
The following packages will be REMOVED:
  ibus-keyman
0 upgraded, 0 newly installed, 1 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 73,7 kB disk space will be freed.
Do you want to continue? [Y/n] 
(Reading database ... 203968 files and directories currently installed.)
Removing ibus-keyman (13.0.103-1~sil1~bionic) ...
Processing triggers for desktop-file-utils (0.23-1ubuntu3.18.04.2) ...
Processing triggers for libglib2.0-0:amd64 (2.56.4-0ubuntu0.18.04.6) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
Processing triggers for shared-mime-info (1.9-2) ...
Unknown media type in type 'all/all'
Unknown media type in type 'all/allfiles'
Processing triggers for gnome-menus (3.13.3-11ubuntu1.1) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...
Processing triggers for mime-support (3.60ubuntu1) ...
```
 
> This will remove the broken packages.

It did.

> Afterwards you can reinstall keyman (or whatever).

I changed to Plasma KDE again&mdash;out of curiosity and because my linguist friend actually uses Kubuntu with KDE.

```
daniel@daniel-Standard-PC-Q35-ICH9-2009:~$ ps -C ibus-daemon -f          
UID        PID  PPID  C STIME TTY          TIME CMD
gdm       1578  1544  0 00:17 tty1     00:00:00 ibus-daemon --xim --panel disable
daniel@daniel-Standard-PC-Q35-ICH9-2009:~$ sudo apt install keyman
[sudo] password for daniel: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  ibus-keyman python3-keyman-config
The following NEW packages will be installed:
  ibus-keyman keyman python3-keyman-config
0 upgraded, 3 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/99,4 kB of archives.
After this operation, 366 kB of additional disk space will be used.
Do you want to continue? [Y/n] 
Selecting previously unselected package ibus-keyman.
(Reading database ... 203969 files and directories currently installed.)
Preparing to unpack .../ibus-keyman_13.0.103-1~sil1~bionic_amd64.deb ...
Unpacking ibus-keyman (13.0.103-1~sil1~bionic) ...
Selecting previously unselected package python3-keyman-config.
Preparing to unpack .../python3-keyman-config_13.0.103-1~sil1~bionic_all.deb ...
Unpacking python3-keyman-config (13.0.103-1~sil1~bionic) ...
Selecting previously unselected package keyman.
Preparing to unpack .../keyman_13.0.103-1~sil1~bionic_all.deb ...
Unpacking keyman (13.0.103-1~sil1~bionic) ...
Setting up ibus-keyman (13.0.103-1~sil1~bionic) ...
Can't connect to IBus.
dpkg: error processing package ibus-keyman (--configure):
 installed ibus-keyman package post-installation script subprocess returned error exit status 1
dpkg: dependency problems prevent configuration of python3-keyman-config:
 python3-keyman-config depends on ibus-keyman; however:
  Package ibus-keyman is not configured yet.

dpkg: error processing package python3-keyman-config (--configure):
 dependency problems - leaving unconfigured
dpkg: dependency problems prevent configuration of keyman:
 keyman depends on python3-keyman-config; however:
  Package python3-keyman-config is not configured yet.

dpkg: error processing package keyman (--configure):                                                                         
 dependency problems - leaving unconfigured                                                                                  
No apport report written because the error message indicates its a followup error from a previous failure.                   
                                                                                                          No apport report written because the error message indicates its a followup error from a previous failure.                                      
                                                                                       Processing triggers for shared-mime-info (1.9-2) ...                                                                                                               
Unknown media type in type 'all/all'                                                                                         
Unknown media type in type 'all/allfiles'
Processing triggers for gnome-menus (3.13.3-11ubuntu1.1) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...
Processing triggers for mime-support (3.60ubuntu1) ...
Processing triggers for desktop-file-utils (0.23-1ubuntu3.18.04.2) ...
Processing triggers for libglib2.0-0:amd64 (2.56.4-0ubuntu0.18.04.6) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
Errors were encountered while processing:
 ibus-keyman
 python3-keyman-config
 keyman
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

I changed back to Ubuntu (Gnome) and could successfully complete the installation and run the keyman configuration utility.

* Can we fix the ibus setup for KDE?
* Can I provide any more data for you?
* My linguist friend might know but I have no clue yet: How to actually use Keyman when it works (like under Gnome already)? I am no Keyman user but I'd just like to enter some chars for testing. Can you give me a hint?

**Comment by ermshiperete** [comment]
Great that these steps sorted the problem! I'll try to look into KDE next week.

For now you can probably manually start ibus by running something like `ibus-daemon --xim &`.

[This](https://help.keyman.com/products/linux/13.0/guide/installing-keyboard) page shows you how to install a keyboard.

**Comment by ermshiperete** [comment]
Uninstall fixed in 13.0.104

**Comment by ermshiperete** [comment]
For using Keyman with KDE, see https://community.software.sil.org/t/howto-using-keyman-for-linux-with-kde
